dali/tensorflow/dali_dataset.cc

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorflow/core/public/version.h"
+
+#if TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION == 12
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wreorder"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#include "tensorflow/core/framework/common_shape_fns.h"
+#include "tensorflow/core/framework/dataset.h"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/partial_tensor_shape.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/tensor.h"
+#pragma GCC diagnostic pop
+#pragma GCC diagnostic pop
+
+#include "dali/c_api/c_api.h"
+
+namespace tensorflow {
+
+namespace data {
+
+namespace {
+
+// TODO extract this guy. By now C&P from daliop
+TensorShape DaliToShape(int64_t* ns) {
+  TensorShape ts;
+  for (int i = 0; ns[i] != 0; ++i)
+    ts.InsertDim(i, ns[i]);
+  delete ns;
+  return ts;
+}
+
+class DALIDatasetOp : public DatasetOpKernel {
+ public:
+  using DatasetOpKernel::DatasetOpKernel;
+
+  DALIDatasetOp(OpKernelConstruction* context)
+  : DatasetOpKernel(context) {
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("serialized_pipeline", &serialized_pipeline_));
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("shapes", &shapes_));
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("dtypes", &dtypes_));
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("prefetch_queue_depth", &prefetch_queue_depth_));
+      OP_REQUIRES_OK(context,
+                   context->GetAttr("num_threads", &num_threads_));
+
+    std::vector<int> devices;
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("devices", &devices));
+
+    pipe_handles_.reserve(devices.size());
+
+    for (size_t  i = 0; i < devices.size(); i++) {
+      pipe_handles_.emplace_back(daliPipelineHandle());
+
+      daliCreatePipeline(&pipe_handles_.back(),
+                  serialized_pipeline_.c_str(),
+                  serialized_pipeline_.length(),
+                  128, // TODO: get batch size from shapes
+                  num_threads_,
+                  devices[i],
+                  prefetch_queue_depth_);
+
+      for (int j = 0; j < prefetch_queue_depth_; j++) {
+        daliRun(&pipe_handles_.back());
```

Comment:
Shouldn't it be:
```suggestion
        daliRun(&pipe_handles_[j]);
```

Reply:
Or simply what you do at L90 :+1:

Reply:
yeah, I think here I should be calling run over the whole collection of pipelines in order to have them all prefetch. Good catch